## Cosine Similarity vs Euclidean distance


5 Articles from each of following categories : Marvel movies, Actor, Data Science, Football players

### Assumption :
- Different catogories differ from each other
- There is similarity within each article Ex. Maximum similarity between the Football players, but minimum similarity between football players and articles from Data Scince because they dont have any connection.

Lets find out either euclidean distance or cosine similarity gives better output

In [1]:
# Marvel movies
b1 = 'Hulk'
b2 = 'X-men'
b3 = 'Spider-man'
b4 = 'Iron man 3'
b5 = 'Deadpool'

# Actors
b6 = 'Hugh Michael Jackman'
b7 = 'Akshay Kumar'
b8 = 'Ashok Saraf'
b9 = 'Nana Patekar'
b10 = 'Joseph Morgan'

# Data Science
b11 = 'Bayesian probability'
b12 = 'Estimation'
b13 = 'Linear regression'
b14 = 'Convolutional neural network'
b15 = 'Analytics'

# Football players
b16 = 'Cristiano Ronaldo'
b17 = 'Lionel Messi'
b18 = 'David Beckham'
b19 = 'David de Gea'
b20 = 'Wayne Rooney'

In [2]:
# Importing require lib
import pandas as pd
import numpy as np
import os
import wikipedia

Function for Euclidean distance

In [3]:
def euclidean_distance(x, y):   
    return np.sqrt(np.sum((x - y) ** 2))

Function for Cosine Similarity 

In [4]:
def cosine_similarity(x, y):
    return np.dot(x, y) / (np.sqrt(np.dot(x, x)) * np.sqrt(np.dot(y, y)))

Importing these articles from wikipedia

In [5]:
# Marvel movies
a1 = wikipedia.page(b1)
a2 = wikipedia.page(b2)
a3 = wikipedia.page(b3)
a4 = wikipedia.page(b4)
a5 = wikipedia.page(b5)

In [6]:
# Actors
a6 = wikipedia.page(b6)
a7 = wikipedia.page(b7)
a8 = wikipedia.page(b8)
a9 = wikipedia.page(b9)
a10 = wikipedia.page(b10)

In [7]:
# Data science
a11 = wikipedia.page(b11)
a12 = wikipedia.page(b12)
a13 = wikipedia.page(b13)
a14 = wikipedia.page(b14)
a15 = wikipedia.page(b15)

In [8]:
# Football players
a16 = wikipedia.page(b16)
a17 = wikipedia.page(b17)
a18 = wikipedia.page(b18)
a19 = wikipedia.page(b19)
a20 = wikipedia.page(b20)

In [9]:
from sklearn.feature_extraction.text import CountVectorizer
# Vectorizing

cv = CountVectorizer()
X = np.array(cv.fit_transform([a1.content, a2.content, a3.content, a4.content, a5.content,
                              a6.content, a7.content, a8.content, a9.content, a10.content,
                              a11.content, a12.content, a13.content, a14.content, a15.content,
                              a16.content, a17.content, a18.content, a19.content, a20.content]).todense())

In [10]:
#Finding length of each article
print(b1, len(a1.content.split()), "\n",
      b2, len(a2.content.split()), "\n",
      b3, len(a3.content.split()), "\n",
      b4, len(a4.content.split()), "\n",
      b5, len(a5.content.split()), "\n",
      b6, len(a6.content.split()), "\n",
      b7, len(a7.content.split()), "\n",
      b8, len(a8.content.split()), "\n",
      b9, len(a9.content.split()), "\n",
      b10, len(a10.content.split()), "\n",
      b11, len(a11.content.split()), "\n",
      b12, len(a12.content.split()), "\n",
      b13, len(a13.content.split()), "\n",
      b14, len(a14.content.split()), "\n",
      b15, len(a15.content.split()), "\n",
      b16, len(a16.content.split()), "\n",
      b17, len(a17.content.split()), "\n",
      b18, len(a18.content.split()), "\n",
      b19, len(a19.content.split()), "\n",
      b20, len(a20.content.split()))

Hulk 8481 
 X-men 7168 
 Spider-man 9180 
 Iron man 3 10469 
 Deadpool 498 
 Hugh Michael Jackman 4203 
 Akshay Kumar 5375 
 Ashok Saraf 796 
 Nana Patekar 1688 
 Joseph Morgan 363 
 Bayesian probability 2363 
 Estimation 906 
 Linear regression 6199 
 Convolutional neural network 8073 
 Analytics 2045 
 Cristiano Ronaldo 13496 
 Lionel Messi 17840 
 David Beckham 14417 
 David de Gea 5203 
 Wayne Rooney 10475


Final Calculating Euclidean distance and Cosine similarity for all articles

In [11]:

list_c_1,list_c_2,list_c_3,list_c_4,list_c_5,list_c_6,list_c_7,list_c_8,list_c_9,list_c_10,list_c_11,list_c_12,list_c_13,list_c_14,list_c_15,list_c_16,list_c_17,list_c_18,list_c_19,list_c_20 = ([] for c in range(20))
list_d_1,list_d_2,list_d_3,list_d_4,list_d_5,list_d_6,list_d_7,list_d_8,list_d_9,list_d_10,list_d_11,list_d_12,list_d_13,list_d_14,list_d_15,list_d_16,list_d_17,list_d_18,list_d_19,list_d_20 = ([] for d in range(20))
list_c = [list_c_1,list_c_2,list_c_3,list_c_4,list_c_5,list_c_6,list_c_7,list_c_8,list_c_9,list_c_10,list_c_11,list_c_12,list_c_13,list_c_14,list_c_15,list_c_16,list_c_17,list_c_18,list_c_19,list_c_20]
list_d = [list_d_1,list_d_2,list_d_3,list_d_4,list_d_5,list_d_6,list_d_7,list_d_8,list_d_9,list_d_10,list_d_11,list_d_12,list_d_13,list_d_14,list_d_15,list_d_16,list_d_17,list_d_18,list_d_19,list_d_20]
f = 0
for (x,y) in zip(list_c,list_d):
    j = 0
    for p in range(20):
        c = round(cosine_similarity(X[f],X[j]),2)
        d = round(euclidean_distance(X[f],X[j]),2)
        x.append(c)
        y.append(d)
        j+=1
        
    f += 1
    

In [12]:
art = [b1,b2,b3,b4,b5,b6,b7,b8,b9,b10,b11,b12,b13,b14,b15,b16,b17,b18,b19,b20]  # Names of article

Creating dataframes for both cosine similarity and euclidean distance

In [13]:
final_cosine = pd.DataFrame(
{'Article':art,
 b1:list_c_1,b2:list_c_2,b3:list_c_3,b4:list_c_4,b5:list_c_5,b6:list_c_6,b7:list_c_7,b8:list_c_8,b9:list_c_9,b10:list_c_10,
b11:list_c_11,b12:list_c_12,b13:list_c_13,b14:list_c_14,b15:list_c_15,b16:list_c_16,b17:list_c_17,b18:list_c_18,b19:list_c_19,b20:list_c_20})

final_euclidean = pd.DataFrame(
{'Article':art,
 b1:list_d_1,b2:list_d_2,b3:list_d_3,b4:list_d_4,b5:list_d_5,b6:list_d_6,b7:list_d_7,b8:list_d_8,b9:list_d_9,b10:list_d_10,
b11:list_d_11,b12:list_d_12,b13:list_d_13,b14:list_d_14,b15:list_d_15,b16:list_d_16,b17:list_d_17,b18:list_d_18,b19:list_d_19,b20:list_d_20})


In [14]:
final_cosine.head()

Article  Hulk  X-men  Spider-man  Iron man 3  Deadpool  \
0        Hulk  1.00   0.80        0.81        0.84      0.69   
1       X-men  0.80   1.00        0.79        0.84      0.64   
2  Spider-man  0.81   0.79        1.00        0.86      0.68   
3  Iron man 3  0.84   0.84        0.86        1.00      0.70   
4    Deadpool  0.69   0.64        0.68        0.70      1.00   

   Hugh Michael Jackman  Akshay Kumar  Ashok Saraf  Nana Patekar  ...  \
0                  0.81          0.77         0.57          0.70  ...   
1                  0.78          0.75         0.55          0.68  ...   
2                  0.81          0.77         0.59          0.71  ...   
3                  0.82          0.79         0.59          0.72  ...   
4                  0.74          0.70         0.56          0.66  ...   

   Bayesian probability  Estimation  Linear regression  \
0                  0.73        0.72               0.76   
1                  0.70        0.71               0.71   
2                  0.72        0.69               0.73   
3                  0.74        0.73               0.76   
4                  0.68        0.64               0.69   

   Convolutional neural network  Analytics  Cristiano Ronaldo  Lionel Messi  \
0                          0.81       0.71               0.80          0.81   
1                          0.77       0.71               0.76          0.78   
2                          0.79       0.69               0.79          0.80   
3                          0.82       0.73               0.81          0.82   
4                          0.74       0.63               0.73          0.73   

   David Beckham  David de Gea  Wayne Rooney  
0           0.82          0.76          0.75  
1           0.79          0.73          0.72  
2           0.80          0.74          0.74  
3           0.82          0.76          0.76  
4           0.73          0.67          0.70  

[5 rows x 21 columns]

In [15]:
final_euclidean.head()

Article    Hulk   X-men  Spider-man  Iron man 3  Deadpool  \
0        Hulk    0.00  566.74      579.52      551.44    905.37   
1       X-men  566.74    0.00      584.11      565.44    678.00   
2  Spider-man  579.52  584.11        0.00      517.02    910.07   
3  Iron man 3  551.44  565.44      517.02        0.00    976.23   
4    Deadpool  905.37  678.00      910.07      976.23      0.00   

   Hugh Michael Jackman  Akshay Kumar  Ashok Saraf  Nana Patekar  ...  \
0                651.84        637.93       897.02        821.65  ...   
1                463.46        468.40       669.72        599.63  ...   
2                654.74        640.32       900.11        823.56  ...   
3                709.19        683.63       966.95        887.37  ...   
4                380.32        479.94        65.48        156.32  ...   

   Bayesian probability  Estimation  Linear regression  \
0                795.37      868.97             626.62   
1                578.58      641.65             509.79   
2                802.38      876.76             655.88   
3                862.19      939.26             683.89   
4                186.97       79.33             549.00   

   Convolutional neural network  Analytics  Cristiano Ronaldo  Lionel Messi  \
0                        545.08     808.76             828.31       1140.46   
1                        503.26     583.30             931.34       1280.59   
2                        581.02     818.38             838.94       1156.19   
3                        582.39     875.59             795.56       1089.82   
4                        738.13     174.55            1316.66       1721.34   

   David Beckham  David de Gea  Wayne Rooney  
0         810.28        649.69        689.58  
1         928.54        481.98        693.68  
2         834.78        661.58        699.53  
3         792.32        705.76        696.32  
4        1339.30        469.19        971.09  

[5 rows x 21 columns]

## Final Cosine Similarity between each article
(Yellow to Dark blue - Maximum to Minimum)

In [16]:
import seaborn as sns
 
cm = sns.light_palette("red", as_cmap=True)
 
final_cosine.style.background_gradient(cmap='viridis')

## Final Euclidean Distance between each article
(Yellow to Dark blue - Maximum to Minimum)

In [17]:
import seaborn as sns
 
cm = sns.light_palette("red", as_cmap=True)
 
final_euclidean.style.background_gradient(cmap='viridis')

### Conclusion : Few highlights
- As we know that euclidean distance uses the length to do the calculation, observe from euclidean distance table the row 16 i.e Lionel Messi is through out yellow which is not in case of cosine similarity. This is because the length of Lionel Messi is highest among all the articles i.e Lionel Messi 17840 (calculted above).

- From cosine similarity table, observe right bottom corner is yellowish this is because that right bottom matrix belong to football players and there is maximum similarity within footfall player article compare other articles of different categories

- Row 7 i.e Ashok Saraf from cosine similarity table is through out dark i.e is because Ashok Saraf is a marathi movie actor and its content doesn't matches with other articles which make sense. 

- Last 5 columns from euclidean distance table are more yellowish, here more yellow implies that the length of these articles are quite similar hence more yellowish